filestorage.go

```diff
@@ -118,191 +120,221 @@ func (fs FileStorage) Stat(key string) (KeyInfo, error) {
 	if ok {
-		// lock already created within process, let caller wait on it
-		return fw, nil
+		// this process is already waiting for the
+		// lock; wait on the same wait chan to avoid
+		// more polling than is necessary
+		fw.wait()
 	}
 
-	// attempt to persist lock to disk by creating lock file
-
-	// parent dir must exist
-	lockDir := fs.lockDir()
-	if err := os.MkdirAll(lockDir, 0700); err != nil {
-		return nil, err
-	}
+	return fs.getLock(key)
+}
 
-	fw = &fileStorageWaiter{
-		key:      key,
-		filename: filepath.Join(lockDir, StorageKeys.safe(key)+".lock"),
-		wg:       new(sync.WaitGroup),
-	}
+func (fs *FileStorage) getLock(key string) error {
+	start := time.Now()
+	filename := fs.lockFilename(key)
 
-	var checkedStaleLock bool // sentinel value to avoid infinite goto-ing
+	for {
+		err := createLockfile(filename)
+		if err == nil {
+			// got the lock, yay
+			return nil
+		}
+		if !os.IsExist(err) {
+			// unexpected error
+			return fmt.Errorf("creating lock file: %v", err)
+		}
 
-createLock:
-	// create the file in a special mode such that an
-	// error is returned if it already exists
-	lf, err := os.OpenFile(fw.filename, os.O_CREATE|os.O_EXCL, 0644)
-	if err != nil {
-		if os.IsExist(err) {
-			// another process has the lock
-
-			// check to see if the lock is stale, if we haven't already
-			if !checkedStaleLock {
-				checkedStaleLock = true
-				if fs.lockFileStale(fw.filename) {
-					log.Printf("[INFO][%s] Lock for '%s' is stale; removing then retrying: %s",
-						fs, key, fw.filename)
-					os.Remove(fw.filename)
-					goto createLock
+		// lock file already exists
+
+		info, err := os.Stat(filename)
+		switch {
+		case os.IsNotExist(err):
+			// must have just been removed; try again to create it
+			continue
+
+		case err != nil:
+			// unexpected error
+			return fmt.Errorf("accessing lock file: %v", err)
+
+		case fileLockIsStale(info):
+			// lock file is stale - delete it and try again to create one
+			log.Printf("[INFO][%s] Lock for '%s' is stale; removing then retrying: %s",
+				fs, key, filename)
+			removeLockfile(filename)
+			continue
+
+		case time.Since(start) > staleLockDuration*2:
+			// should never happen, hopefully
+			return fmt.Errorf("possible deadlock: %s passed trying to obtain lock for %s",
+				time.Since(start), key)
+
+		default:
+			// lock file exists and is not stale; wait on it,
+			// and allow other threads in this process to share
+			// our wait logic so we don't excessively poll
+			fs.lockWaitersMu.Lock()
+			fw, ok := fs.lockWaiters[key]
+			if ok {
+				// waiter already exists for this lock; use
+				// it, then try again to obtain the lock
+				fs.lockWaitersMu.Unlock()
+				fw.wait()
+				continue
+			}
+			waitChan := make(chan struct{})
+			fs.lockWaiters[key] = waitChan
+			fs.lockWaitersMu.Unlock()
+
+			// poll until the lock is available; ours should be
+			// the only thread in this process polling the disk
+			for time.Since(start) < staleLockDuration {
```

Comment:
I think there's a small possibility that this condition can be false before we enter the loop and we'd fail to close `waitChan`.